File: lib/tts.ts
// Cartesia-only TTS passthrough used by `/api/tts/stream` and the
// combined `/api/speak` route.
//
// No fallback providers by design — the old Cartesia → Fish → OpenAI
// cascade was what caused the "long dead air / no audio at all"
// behaviour: an aggressive Cartesia timeout would abort a request that
// was about to succeed, bounce through Fish (usually unconfigured), and
// finally land on OpenAI tts-1/nova for another ~1–2s of silence.
// Timeout is generous (120s) so a slow-but-real response wins instead
// of being aborted.

const CARTESIA_TTS_URL = "https://api.cartesia.ai/tts/bytes";
const CARTESIA_VERSION = "2024-11-13";
const CARTESIA_TTFB_TIMEOUT_MS = 120_000;

export type TtsBackend = "cartesia";

export type StreamTtsOptions = {
  text: string;
  voiceId?: string;
  lang?: "en" | "zh";
  turnId?: string;
  emotion?: string[];
  speed?: string | number | null;
};

export type StreamTtsResult = {
  stream: ReadableStream<Uint8Array>;
  backend: TtsBackend;
};

const VALID_EMOTIONS = new Set([
  "anger",
  "positivity",
  "surprise",
  "sadness",
  "curiosity",
]);
const VALID_INTENSITIES = new Set(["lowest", "low", "high", "highest"]);
const VALID_SPEEDS = new Set([
  "slowest",
  "slow",
  "normal",
  "fast",
  "fastest",
]);

export function sanitizeEmotion(raw: unknown): string[] {
  if (!Array.isArray(raw)) return [];
  const out: string[] = [];
  for (const item of raw) {
    if (typeof item !== "string") continue;
    const trimmed = item.trim().toLowerCase();
    if (!trimmed) continue;
    const [name, intensity] = trimmed.split(":");
    if (!VALID_EMOTIONS.has(name)) continue;
    if (intensity && !VALID_INTENSITIES.has(intensity)) continue;
    out.push(intensity ? `${name}:${intensity}` : name);
    if (out.length >= 3) break;
  }
  return out;
}

export function sanitizeSpeed(raw: unknown): string | number | null {
  if (typeof raw === "number" && Number.isFinite(raw)) {
    return Math.max(-1, Math.min(1, raw));
  }
  if (typeof raw === "string") {
    const trimmed = raw.trim().toLowerCase();
    if (VALID_SPEEDS.has(trimmed)) return trimmed;
  }
  return null;
}

// Some Cerebras Llama replies come back as UTF-8 bytes *interpreted* as
// latin-1 — e.g. `è¾²ä¹°è¿ç­é¼ ææ­¥` instead of the actual Chinese it
// meant to produce. When that string is piped to Cartesia, the voice
// pronounces the garbled latin characters verbatim (gibberish). This
// detects the pattern (every char sitting in the 0x80–0xFF range, no
// codepoints above) and, if the re-byte-then-decode-as-UTF-8 round trip
// yields real CJK, returns the recovered text. On any ambiguity we
// leave the original alone.
function fixMojibake(s: string): string {
  if (!s) return s;
  let suspect = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c > 0xff) return s;
    if (c >= 0x80) suspect++;
  }
  if (suspect < 3 || suspect < s.length * 0.3) return s;
  try {
    const bytes = new Uint8Array(s.length);
    for (let i = 0; i < s.length; i++) bytes[i] = s.charCodeAt(i);
    const decoded = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    if (/[\u3040-\u30ff\u3400-\u9fff\uac00-\ud7af]/.test(decoded)) {
      // eslint-disable-next-line no-console
      console.log(
        `[tts] ⚠ mojibake recovered: ${s.slice(0, 40)}… → ${decoded.slice(0, 40)}…`
      );
      return decoded;
    }
  } catch {
    // Not valid UTF-8 — leave it.
  }
  return s;
}

// Returns the Cartesia audio stream, or `null` when the key is missing
// or Cartesia returned a non-audio response. The caller 503s in that
// case — there is no secondary provider anymore.
export async function streamTts(
  opts: StreamTtsOptions
): Promise<StreamTtsResult | null> {
  const text = fixMojibake(opts.text).trim().slice(0, 600);
  if (!text) return null;
  const voiceId = (opts.voiceId ?? "").trim();
  const lang = opts.lang === "zh" ? "zh" : "en";
  const turnId = (opts.turnId ?? "?").slice(0, 16) || "?";
  const tag = ` #${turnId}`;
  const emotion = opts.emotion ?? [];
  const speed = opts.speed ?? null;

  const cartesiaKey = process.env.CARTESIA_API_KEY?.trim();
  if (!cartesiaKey) {
    // eslint-disable-next-line no-console
    console.log(`[tts cartesia${tag}] ✖ CARTESIA_API_KEY not set`);
    return null;
  }

  const modelId = process.env.CARTESIA_MODEL_ID?.trim() || "sonic-3";
  const defaultVoice =
    lang === "zh"
      ? process.env.CARTESIA_VOICE_ID_ZH?.trim() ||
        "0cd0cde2-3b93-42b5-bcb9-f214a591aa29"
      : process.env.CARTESIA_VOICE_ID_EN?.trim() ||
        "a0e99841-438c-4a64-b679-ae501e7d6091";
  const uuidRe =
    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  const chosenVoice = uuidRe.test(voiceId) ? voiceId : defaultVoice;
  type CartesiaVoice = {
    mode: "id";
    id: string;
    experimental_controls?: {
      emotion?: string[];
      speed?: string | number;
    };
  };
  const voicePayload: CartesiaVoice = { mode: "id", id: chosenVoice };
  if (emotion.length > 0 || speed !== null) {
    voicePayload.experimental_controls = {
      ...(emotion.length > 0 ? { emotion } : {}),
      ...(speed !== null ? { speed } : {}),
    };
  }
  const t0 = Date.now();
  // eslint-disable-next-line no-console
  console.log(
    `[tts cartesia${tag}] → model=${modelId} voice=${chosenVoice} lang=${lang} text=${text.length}ch emotion=[${emotion.join(",") || "-"}] speed=${speed ?? "-"}`
  );
  const ctrl = new AbortController();
  const timer = setTimeout(
    () =>
      ctrl.abort(
        new Error(`cartesia ttfb > ${CARTESIA_TTFB_TIMEOUT_MS}ms`)
      ),
    CARTESIA_TTFB_TIMEOUT_MS
  );
  try {
    const res = await fetch(CARTESIA_TTS_URL, {
      method: "POST",
      headers: {
        "X-API-Key": cartesiaKey,
        "Cartesia-Version": CARTESIA_VERSION,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model_id: modelId,
        transcript: text,
        voice: voicePayload,
        output_format: {
          container: "mp3",
          bit_rate: 128000,
          sample_rate: 44100,
        },
        language: lang === "zh" ? "zh" : "en",
      }),
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    const ct = res.headers.get("content-type") ?? "";
    if (res.ok && res.body && !ct.includes("application/json")) {
      // eslint-disable-next-line no-console
      console.log(
        `[tts cartesia${tag}] ✓ ttfb=${Date.now() - t0}ms streaming audio/mpeg`
      );
      return { stream: res.body, backend: "cartesia" };
    }
    const err = await res.text().catch(() => "");
    // eslint-disable-next-line no-console
    console.log(
      `[tts cartesia${tag}] ✖ ${res.status} in ${Date.now() - t0}ms: ${err.slice(0, 200)}`
    );
    return null;
  } catch (err) {
    clearTimeout(timer);
    // eslint-disable-next-line no-console
    console.log(
      `[tts cartesia${tag}] ✖ exception after ${Date.now() - t0}ms: ${err instanceof Error ? err.message : String(err)}`
    );
    return null;
  }
}

File: backend/src/omni_backend/services/tts.py
"""Cartesia-only TTS passthrough used by ``/api/tts/stream`` and the
combined ``/api/speak`` route.

No fallback providers by design — the old Cartesia → Fish → OpenAI
cascade was what caused the "long dead air / no audio at all"
behaviour: an aggressive Cartesia timeout would abort a request that
was about to succeed, bounce through Fish (usually unconfigured), and
finally land on OpenAI tts-1/nova for another ~1–2s of silence.
Timeout is generous (120s) so a slow-but-real response wins instead
of being aborted.
"""

from __future__ import annotations

import logging
import os
import re
import time
from dataclasses import dataclass
from typing import Any, AsyncIterator, Iterable, Literal, Optional, Union

import httpx

logger = logging.getLogger(__name__)

CARTESIA_TTS_URL = "https://api.cartesia.ai/tts/bytes"
CARTESIA_VERSION = "2024-11-13"
CARTESIA_TTFB_TIMEOUT_MS = 120_000

TtsBackend = Literal["cartesia"]

VALID_EMOTIONS = {"anger", "positivity", "surprise", "sadness", "curiosity"}
VALID_INTENSITIES = {"lowest", "low", "high", "highest"}
VALID_SPEEDS = {"slowest", "slow", "normal", "fast", "fastest"}

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)
_CJK_RE = re.compile(r"[\u3040-\u30ff\u3400-\u9fff\uac00-\ud7af]")


@dataclass
class StreamTtsResult:
    stream: AsyncIterator[bytes]
    backend: TtsBackend


def sanitize_emotion(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        if not isinstance(item, str):
            continue
        trimmed = item.strip().lower()
        if not trimmed:
            continue
        parts = trimmed.split(":", 1)
        name = parts[0]
        intensity = parts[1] if len(parts) > 1 else ""
        if name not in VALID_EMOTIONS:
            continue
        if intensity and intensity not in VALID_INTENSITIES:
            continue
        out.append(f"{name}:{intensity}" if intensity else name)
        if len(out) >= 3:
            break
    return out


def sanitize_speed(raw: Any) -> Union[str, float, None]:
    if isinstance(raw, bool):
        # JS `typeof true === "boolean"`, not "number" — exclude explicitly.
        return None
    if isinstance(raw, (int, float)):
        value = float(raw)
        if value != value or value in (float("inf"), float("-inf")):
            return None
        return max(-1.0, min(1.0, value))
    if isinstance(raw, str):
        trimmed = raw.strip().lower()
        if trimmed in VALID_SPEEDS:
            return trimmed
    return None


def fix_mojibake(s: str) -> str:
    """Recover UTF-8 text that was double-decoded as latin-1.

    Detects every char sitting in the 0x80–0xFF range (no codepoints
    above) and, if the re-byte-then-decode-as-UTF-8 round trip yields
    real CJK, returns the recovered text. On any ambiguity we leave
    the original alone.
    """
    if not s:
        return s
    suspect = 0
    for ch in s:
        c = ord(ch)
        if c > 0xFF:
            return s
        if c >= 0x80:
            suspect += 1
    if suspect < 3 or suspect < len(s) * 0.3:
        return s
    try:
        raw_bytes = bytes(ord(ch) for ch in s)
        decoded = raw_bytes.decode("utf-8")
        if _CJK_RE.search(decoded):
            logger.info(
                "[tts] \u26a0 mojibake recovered: %s\u2026 \u2192 %s\u2026",
                s[:40],
                decoded[:40],
            )
            return decoded
    except (UnicodeDecodeError, ValueError):
        # Not valid UTF-8 — leave it.
        pass
    return s


async def stream_tts(
    text: str,
    voice_id: Optional[str] = None,
    lang: Optional[str] = None,
    turn_id: Optional[str] = None,
    emotion: Optional[Iterable[str]] = None,
    speed: Union[str, float, None] = None,
) -> Optional[StreamTtsResult]:
    """Return the Cartesia audio stream, or ``None`` when the key is
    missing or Cartesia returned a non-audio response. The caller 503s
    in that case — there is no secondary provider anymore.
    """
    text = fix_mojibake(text or "").strip()[:600]
    if not text:
        return None
    voice_id_s = (voice_id or "").strip()
    lang_s = "zh" if lang == "zh" else "en"
    turn = (turn_id or "?")[:16] or "?"
    tag = f" #{turn}"
    emotion_list = list(emotion) if emotion else []

    cartesia_key_raw = os.environ.get("CARTESIA_API_KEY")
    cartesia_key = cartesia_key_raw.strip() if cartesia_key_raw else ""
    if not cartesia_key:
        logger.info("[tts cartesia%s] \u2716 CARTESIA_API_KEY not set", tag)
        return None

    model_id_raw = os.environ.get("CARTESIA_MODEL_ID")
    model_id = (model_id_raw.strip() if model_id_raw else "") or "sonic-3"

    if lang_s == "zh":
        voice_env = os.environ.get("CARTESIA_VOICE_ID_ZH")
        default_voice = (
            voice_env.strip() if voice_env else ""
        ) or "0cd0cde2-3b93-42b5-bcb9-f214a591aa29"
    else:
        voice_env = os.environ.get("CARTESIA_VOICE_ID_EN")
        default_voice = (
            voice_env.strip() if voice_env else ""
        ) or "a0e99841-438c-4a64-b679-ae501e7d6091"
    chosen_voice = voice_id_s if _UUID_RE.match(voice_id_s) else default_voice

    voice_payload: dict[str, Any] = {"mode": "id", "id": chosen_voice}
    if emotion_list or speed is not None:
        controls: dict[str, Any] = {}
        if emotion_list:
            controls["emotion"] = emotion_list
        if speed is not None:
            controls["speed"] = speed
        voice_payload["experimental_controls"] = controls

    t0 = time.monotonic()
    logger.info(
        "[tts cartesia%s] \u2192 model=%s voice=%s lang=%s text=%dch emotion=[%s] speed=%s",
        tag,
        model_id,
        chosen_voice,
        lang_s,
        len(text),
        ",".join(emotion_list) if emotion_list else "-",
        speed if speed is not None else "-",
    )

    body = {
        "model_id": model_id,
        "transcript": text,
        "voice": voice_payload,
        "output_format": {
            "container": "mp3",
            "bit_rate": 128000,
            "sample_rate": 44100,
        },
        "language": "zh" if lang_s == "zh" else "en",
    }
    headers = {
        "X-API-Key": cartesia_key,
        "Cartesia-Version": CARTESIA_VERSION,
        "Content-Type": "application/json",
    }
    ttfb_s = CARTESIA_TTFB_TIMEOUT_MS / 1000.0
    timeout = httpx.Timeout(ttfb_s, connect=ttfb_s, read=ttfb_s, write=ttfb_s)

    client = httpx.AsyncClient(timeout=timeout)
    try:
        req = client.build_request(
            "POST", CARTESIA_TTS_URL, headers=headers, json=body
        )
        response = await client.send(req, stream=True)
    except Exception as err:  # noqa: BLE001
        elapsed_ms = int((time.monotonic() - t0) * 1000)
        logger.info(
            "[tts cartesia%s] \u2716 exception after %dms: %s",
            tag,
            elapsed_ms,
            str(err),
        )
        await client.aclose()
        return None

    ct = response.headers.get("content-type", "") or ""
    if response.status_code < 400 and "application/json" not in ct:
        ttfb_ms = int((time.monotonic() - t0) * 1000)
        logger.info(
            "[tts cartesia%s] \u2713 ttfb=%dms streaming audio/mpeg", tag, ttfb_ms
        )

        async def _iter() -> AsyncIterator[bytes]:
            try:
                async for chunk in response.aiter_bytes():
                    if chunk:
                        yield chunk
            finally:
                await response.aclose()
                await client.aclose()

        return StreamTtsResult(stream=_iter(), backend="cartesia")

    try:
        err_bytes = await response.aread()
        err_text = err_bytes.decode("utf-8", errors="replace")
    except Exception:  # noqa: BLE001
        err_text = ""
    elapsed_ms = int((time.monotonic() - t0) * 1000)
    logger.info(
        "[tts cartesia%s] \u2716 %d in %dms: %s",
        tag,
        response.status_code,
        elapsed_ms,
        err_text[:200],
    )
    await response.aclose()
    await client.aclose()
    return None
